source/slang/slang-constructor-utility.cpp

```diff
@@ -0,0 +1,264 @@
+// slang-constructor-utility.cpp
+#include "slang-check-impl.h"
+
+namespace Slang
+{
+
+    DefaultConstructExpr* createDefaultConstructExprForType(ASTBuilder* m_astBuilder, QualType type, SourceLoc loc)
+    {
+        auto defaultConstructExpr = m_astBuilder->create<DefaultConstructExpr>();
+        defaultConstructExpr->type = type;
+        defaultConstructExpr->loc = loc;
+        return defaultConstructExpr;
+    }
+
+    ConstructorDecl* _getDefaultCtor(StructDecl* structDecl)
+    {
+        for (auto ctor : structDecl->getMembersOfType<ConstructorDecl>())
+        {
+            if (!ctor->body || ctor->members.getCount() != 0)
+                continue;
+            return ctor;
+        }
+        return nullptr;
+    }
+    bool allParamHaveInitExpr(ConstructorDecl* ctor)
+    {
+        for (auto i : ctor->getParameters())
+            if (!i->initExpr)
+                return false;
+        return true;
+    }
+    List<ConstructorDecl*> _getCtorList(ASTBuilder* m_astBuilder, SemanticsVisitor* visitor, StructDecl* structDecl, ConstructorDecl** defaultCtorOut)
+    {
+        List<ConstructorDecl*> ctorList;
+
+        auto ctorLookupResult = lookUpMember(
+            m_astBuilder,
+            visitor,
+            visitor->getName("$init"),
```

Comment:
We should cache this name on SharedASTBuilder, like we do for the name `This`.

Reply:
addressed here:
https://github.com/shader-slang/slang/pull/5107/commits/54014949789e700a589f5eea098443f3e2bc337c#diff-dad759cd0132d623bee9253550a0bce736d087bdc3157dd1e11e414cb5f6bc48R43